niff-1.02/niffsdk1.02/niffio/doc/man/tags.man
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH tags 3 "niffio" "14/Jun/96" "NIFFIO Documentation"
.IX Title "tags 3"
.UC
.IX Name "tags - tag creation and navigation routines"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
tags \- tag creation and navigation routines
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Sh "Tag Creation"
.IX Subsection "Tag Creation"
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIOTagCreate()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIOTagFinalize()\fR
.Sh "Tag Navigation"
.IX Subsection "Tag Navigation"
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIOTagDescend()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIOTagAscend()\fR
.SH "NIFFIOTagCreate"
.IX Header "NIFFIOTagCreate"
Start a new tag in a NIFF file.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     NIFFIOTagCreate(NIFFIOFile *pnf, NIFFIOTag *ptag)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
T <ptag->tagid> must contain the tag \s-1ID\s0 of the new tag.
.Ip "\(bu" 4
.IX Item "\(bu"
T <ptag->tagsizeData> will be written to the file but
does not need to have a correct value.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
Writes a new tag header to <*pnf>.
.Ip "\(bu" 4
.IX Item "\(bu"
Marks the tag as \*(L"dirty\*(R" (means the data size may need updating).
.Sh "\s-1OBLIGATIONS\s0"
.IX Subsection "\s-1OBLIGATIONS\s0"
\fINIFFIOTagFinalize()\fR must eventually be called with <*ptag>
to finish writing the tag.
If \fINIFFIOTagFinalize()\fR determines that <ptag->tagsizeData> is correct
then it will not seek back in the file to correct it.
.SH "NIFFIOTagFinalize"
.IX Header "NIFFIOTagFinalize"
Finish writing a tag to a NIFFIOFile.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     NIFFIOTagFinalize(NIFFIOFile *pnf, NIFFIOTag *ptag)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
T <*ptag> must be the result of a call to NIFFIOTagCreate.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
Updates the tag data size on the file and in <*ptag> (if necessary).
.Ip "\(bu" 4
.IX Item "\(bu"
Writes a \s-1NUL\s0 pad byte (if necessary).
.Ip "\(bu" 4
.IX Item "\(bu"
Leaves the file positioned at the end of the new tag.
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On failure, the file position is undefined.
.SH "NIFFIOTagDescend"
.IX Header "NIFFIOTagDescend"
Read a tag header from a RIFF file.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     NIFFIOTagDescend(NIFFIOFile *pnf, NIFFIOTag *ptag)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
The file must be positioned at the start of a tag.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
Leaves the file positioned after the tag's size field.
.Ip "\(bu" 4
.IX Item "\(bu"
<ptag->tagid> and
<ptag->tagsizeData> are updated from the
tag's header.
.SH "NIFFIOTagAscend"
.IX Header "NIFFIOTagAscend"
Position a NIFFIOFile after a specified tag.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     NIFFIOTagAscend(NIFFIOFile *pnf, NIFFIOTag *ptag)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
T <*ptag> was returned by NIFFIOTagDescend.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
T <*pnf> is postioned at the end of <*ptag>
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On failure, the file position is undefined.

.rn }` ''


niffsdk1.02/niffio/src/lib/niffio/tags.c
#ifndef lint
static char rcsid[] =
"$Id: tags.c,v 1.1 1996/06/06 04:57:45 tim Exp $";
#endif
/***************************************************************************/
/*
 * NAME
 * ====
 * tags - tag creation and navigation routines
 * 
 * SYNOPSIS
 * ========
 *
 * Tag Creation
 * ------------
 * - NIFFIOTagCreate()
 * - NIFFIOTagFinalize()
 * 
 * Tag Navigation
 * --------------
 * - NIFFIOTagDescend()
 * - NIFFIOTagAscend()
 */
/***************************************************************************/

#include <assert.h>

#include "niffio.h"

#ifndef NDEBUG

/* 
 * The following variables and routines support a debugging feature
 * designed to catch unbalanced calls to NIFFIOTagCreate() and
 * NIFFIOTagFinalize(). 
 */
#include <stdlib.h>
static int _isCreateCalled = 0;         /* true if NIFFIOCreateTag() called
                                           at least once */
static int _CreateFinalizeMatch = 0;    /* counts outstanding calls to
                                           NIFFIOTagFinalize() */
static void _CheckCreateFinalize(void); 

/*
 * Make sure NIFFIOTagCreate() and NIFFIOTagFinalize() are called
 * and equal number of times.
 * This function will be registered with atexit() the first time
 * NIFFIOTagCreate is called.
 */
static void
_CheckCreateFinalize(void)
{
        assert(_CreateFinalizeMatch == 0);
}
#endif

/***************************************************************************/
/*
 * NIFFIOTagCreate
 * ===============
 * Start a new tag in a NIFF file.
 */
RIFFIOSuccess
NIFFIOTagCreate(NIFFIOFile *pnf, NIFFIOTag *ptag)
/*
 * ENTRY
 * -----
 * - T <ptag->tagid> must contain the tag ID of the new tag.
 * - T <ptag->tagsizeData> will be written to the file but 
 *   does not need to have a correct value.
 * 
 * EXIT
 * ----
 * - Writes a new tag header to <*pnf>.
 * - Marks the tag as "dirty" (means the data size may need updating).
 *
 * OBLIGATIONS
 * -----------
 * NIFFIOTagFinalize() must eventually be called with <*ptag> 
 * to finish writing the tag.
 * If NIFFIOTagFinalize() determines that <ptag->tagsizeData> is correct 
 * then it will not seek back in the file to correct it.
 */
/***************************************************************************/
{
    char strModule[] = "NIFFIOTagCreate";
    
    /* Check the arguments */
    assert (pnf != 0);
    assert (ptag != 0);
    
#ifndef NDEBUG
    /*
     * Is this the first time NIFFIOTagCreate was called?
     */
    if (! _isCreateCalled)
    {
        /* yes, initialize the create/finalize balance debugging */
        assert(atexit(_CheckCreateFinalize) == 0);
        
        _isCreateCalled = 1;
    }
    
    /* 
     * Increment the number of times NIFFIOTagCreate was
     * called relative to NIFFIOTagFinalize
     */
    _CreateFinalizeMatch++;
    
#endif
    
    /*
     * Write the tag id
     */
    if (! NIFFIOWrite8(pnf, ptag->tagid))
    {
        RIFFIOError(strModule, "Failed to write tag id %u", ptag->tagid);
        return RIFFIO_FAIL;
    }
    
    /*
     * Write the tag size
     */
    if (! NIFFIOWrite8(pnf, ptag->tagsizeData))
    {
        RIFFIOError(strModule, 
                    "Failed to write tag size %u", ptag->tagsizeData);
        return RIFFIO_FAIL;
    }
    
    /* 
     * Remember the offset of the tag's data
     */
    ptag->offsetData = NIFFIOTell(pnf);
    
    /*
     * Mark the tag as dirty
     */
    ptag->isDirty = 1;
    
    return RIFFIO_OK;
}


/***************************************************************************/
/*
 * NIFFIOTagFinalize
 * =================
 * Finish writing a tag to a NIFFIOFile.
 */
RIFFIOSuccess
NIFFIOTagFinalize(NIFFIOFile *pnf, NIFFIOTag *ptag)
/*
 * ENTRY
 * -----
 * T <*ptag> must be the result of a call to NIFFIOTagCreate.
 * 
 * EXIT
 * ----
 *    - Updates the tag data size on the file and in <*ptag> (if necessary).
 *    - Writes a NUL pad byte (if necessary).
 *    - Leaves the file positioned at the end of the new tag.
 *
 * ERRORS
 * ------
 * On failure, the file position is undefined.
 */
/***************************************************************************/
{
    char strModule[] = "NIFFIOTagFinalize";
    
    RIFFIOOffset offsetTagEnd; /* Offset of end of tag, not including 
                                  possible pad byte */
    BYTE tagsizeNew;          /* Calculated size of a dirty tag */
    
    /* Check the arguments */
    assert (pnf != 0);
    assert (ptag != 0);
    
    /*
     * Only finalize newly created tags
     */
    assert (ptag->isDirty);
    
    
#ifndef NDEBUG
    
    /* 
     * Decrement the number of times RIFFIOChunkCreate was called
     * relative to RIFFIOChunkFinalize
     */
    _CreateFinalizeMatch--;
    
#endif

    offsetTagEnd = NIFFIOTell(pnf);
    
    /*
     * Make sure we aren't trying to update a tag
     * with its end positioned before its beginning
     */
    assert (offsetTagEnd >= ptag->offsetData);
        
    /* 
     * Calculate the tag's size 
     */
    tagsizeNew = offsetTagEnd - ptag->offsetData;
        
    /*
     * Write a nul pad byte if the tag size is an
     * odd number of bytes
     */
    if (tagsizeNew % 2)
    {
        if (! NIFFIOWrite8(pnf, '\0'))
        {
            RIFFIOError(strModule, "Failed to write pad byte at end of tag");
            return RIFFIO_FAIL;
        }
    }
        
    /* 
     * If the newly calculated tag size differs 
     * from its created value, then update the size stored in the file
     */
    if (tagsizeNew != ptag->tagsizeData)
    {
        /* Seek to the size byte */
        if(! NIFFIOSeek(pnf, ptag->offsetData - 1, RIFFIO_SEEK_SET))
        {
            RIFFIOError(strModule, "Failed to seek to tag size");
            return RIFFIO_FAIL;
        }
                
        /* Write the size byte */
        if (! NIFFIOWrite8(pnf, tagsizeNew))
        {
            RIFFIOError(strModule, "Failed to write tag size");
            return RIFFIO_FAIL;
        }

        ptag->tagsizeData = tagsizeNew;         
    }

    /* The tag structure now reflects to correct tag size */
    assert(ptag->tagsizeData == tagsizeNew);

    /* Mark the tag as "clean" */
    ptag->isDirty = 0;

    /*
     * Leave the file positioned after the tag
     */
    if (! NIFFIOTagAscend(pnf, ptag))
    {
        RIFFIOError(strModule, "Failed to ascend past newly created tag");
        return RIFFIO_FAIL;
    }

    return RIFFIO_OK;
        
}

/***************************************************************************/
/*
 * NIFFIOTagDescend
 * ================
 * Read a tag header from a RIFF file.
 */
RIFFIOSuccess
NIFFIOTagDescend(NIFFIOFile *pnf, NIFFIOTag *ptag)
/*
 * ENTRY
 * -----
 * The file must be positioned at the start of a tag.
 * 
 * EXIT
 * ----
 *  - Leaves the file positioned after the tag's size field.
 *  - <ptag->tagid> and 
 *    <ptag->tagsizeData> are updated from the
 *    tag's header.
 */ 
/***************************************************************************/
{
    char strModule[] = "NIFFIOTagDescend";

    /* Check the arguments */
    assert (pnf != 0);
    assert (ptag != 0);

    /* Read the tag's id */
    if (! NIFFIORead8(pnf, &(ptag->tagid)))
    {
        RIFFIOError(strModule, "Failed to read the tag's id");
        return RIFFIO_FAIL;
    }

    /* Read the tag's data size */
    if (! NIFFIORead8(pnf, &(ptag->tagsizeData)))
    {
        RIFFIOError(strModule, "Failed to read the tag's size");
        return RIFFIO_FAIL;
    }
  
    /* Remember the offset of the tag data */
    ptag->offsetData = NIFFIOTell(pnf);

    /* Mark the chunk as clean */
    ptag->isDirty = 0;

    /* 
     * Leave the file positioned at the 
     * beginning of the tag data.
     */
    return RIFFIO_OK;
}

/***************************************************************************/
/*
 * NIFFIOTagAscend
 * ===============
 * Position a NIFFIOFile after a specified tag.
 */
RIFFIOSuccess
NIFFIOTagAscend(NIFFIOFile *pnf, NIFFIOTag *ptag)
/*
 * ENTRY
 * -----
 * T <*ptag> was returned by NIFFIOTagDescend.
 * 
 * EXIT
 * ----
 * T <*pnf> is postioned at the end of <*ptag>
 * 
 * ERRORS
 * ------
 * On failure, the file position is undefined. 
 */
/***************************************************************************/
{
    char strModule[] = "NIFFIOTagAscend";


    /* Check the arguments */
    assert (pnf != 0);
    assert (ptag != 0);

    /*
     * We can't handle tags that haven't been finalized
     */
    assert (ptag->isDirty == 0);

    /*
     * Leave the user positioned after the tag and any
     * nul padding byte
     */
    if (! NIFFIOSeek(pnf,
                     ptag->offsetData 
                     + ptag->tagsizeData 
                     + (ptag->tagsizeData % 2),
                     RIFFIO_SEEK_SET))
    {
        RIFFIOError(strModule, "Failed to seek past of tag");
        return RIFFIO_FAIL; 
    }
                   
    return RIFFIO_OK;

}












